Change to Sun version number

Index: e2fsprogs-1.41.12/version.h
===================================================================
--- e2fsprogs-1.41.12.orig/version.h
+++ e2fsprogs-1.41.12/version.h
@@ -7,5 +7,6 @@
  * file may be redistributed under the GNU Public License v2.
  */
 
-#define E2FSPROGS_VERSION "1.41.12"
-#define E2FSPROGS_DATE "17-May-2010"
+#define E2FSPROGS_VERSION "1.41.12.2.ora1"
+/* Don't forget to update patches/ChangeLog */
+#define E2FSPROGS_DATE "14-Aug-2010"
